File: server/src/engine/win-detector.ts
import { Team, GameResult } from './types';

export function determineWinner(
  finishOrder: string[],
  teams: Map<string, Team>
): GameResult {
  const firstFinisherId = finishOrder[0];
  const lastFinisherId = finishOrder[finishOrder.length - 1];
  const firstTeam = teams.get(firstFinisherId)!;
  const lastTeam = teams.get(lastFinisherId)!;

  const finishOrderResult = finishOrder.map((playerId, index) => ({
    playerId,
    position: index + 1,
  }));

  const redTeam: { playerId: string; isCaught: boolean }[] = [];
  const blackTeam: { playerId: string; isCaught: boolean }[] = [];

  for (const [playerId, team] of teams) {
    const isCaught = playerId === lastFinisherId;
    if (team === 'red') {
      redTeam.push({ playerId, isCaught });
    } else {
      blackTeam.push({ playerId, isCaught });
    }
  }

  let winner: Team | 'draw';
  if (firstTeam !== lastTeam) {
    winner = firstTeam;
  } else {
    winner = 'draw';
  }

  return {
    winner,
    finishOrder: finishOrderResult,
    redTeam,
    blackTeam,
    firstFinisherId,
    caughtPlayerId: winner !== 'draw' ? lastFinisherId : null,
  };
}

export function isGameOver(
  unfinishedCount: number,
  totalPlayers: number
): boolean {
  return unfinishedCount <= 1;
}

File: client/lib/engine/win_detector.dart
import 'types.dart';

GameResult determineWinner(List<String> finishOrder, Map<String, String> teams) {
  final firstFinisherId = finishOrder.first;
  final lastFinisherId = finishOrder.last;
  final firstTeam = teams[firstFinisherId]!;
  final lastTeam = teams[lastFinisherId]!;

  final finishOrderResult = finishOrder.asMap().entries.map((e) => {
    'playerId': e.value,
    'position': e.key + 1,
  }).toList();

  final redTeam = <Map<String, dynamic>>[];
  final blackTeam = <Map<String, dynamic>>[];

  for (final entry in teams.entries) {
    final isCaught = entry.key == lastFinisherId;
    if (entry.value == 'red') {
      redTeam.add({'playerId': entry.key, 'isCaught': isCaught});
    } else {
      blackTeam.add({'playerId': entry.key, 'isCaught': isCaught});
    }
  }

  String winner;
  if (firstTeam != lastTeam) {
    winner = firstTeam;
  } else {
    winner = 'draw';
  }

  return GameResult(
    winner: winner,
    finishOrder: finishOrderResult,
    redTeam: redTeam,
    blackTeam: blackTeam,
    firstFinisherId: firstFinisherId,
    caughtPlayerId: winner != 'draw' ? lastFinisherId : null,
  );
}

bool isGameOver(int unfinishedCount) {
  return unfinishedCount <= 1;
}
